File: app/src/main/java/com/simon/harmonichackernews/UserItemListRepository.java
package com.simon.harmonichackernews;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.simon.harmonichackernews.data.Bookmark;
import com.simon.harmonichackernews.utils.Utils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class UserItemListRepository {
    enum Source {
        FAVORITES,
        UPVOTED
    }

    static class Snapshot {
        final ArrayList<Integer> itemIds;
        final Set<Integer> commentIds;

        Snapshot(ArrayList<Integer> itemIds, Set<Integer> commentIds) {
            this.itemIds = itemIds;
            this.commentIds = commentIds;
        }
    }

    @NonNull
    static Snapshot normalizeSnapshot(@NonNull List<Integer> itemIds,
                                      @NonNull List<Integer> commentIds) {
        ArrayList<Integer> normalizedItemIds = normalizeItemIds(itemIds);
        return new Snapshot(normalizedItemIds, normalizeCommentIds(normalizedItemIds, commentIds));
    }

    @NonNull
    static Snapshot loadCachedSnapshot(@Nullable Context context, @NonNull Source source) {
        ArrayList<Integer> itemIds = new ArrayList<>();
        if (context == null) {
            return new Snapshot(itemIds, new HashSet<>());
        }

        ArrayList<Bookmark> items = loadCache(context, source);
        for (Bookmark item : items) {
            if (!itemIds.contains(item.id)) {
                itemIds.add(item.id);
            }
        }

        Collections.sort(itemIds, (id1, id2) -> Integer.compare(id2, id1));
        return new Snapshot(itemIds, loadCommentIds(context, source));
    }

    @NonNull
    static ArrayList<Bookmark> loadCache(@Nullable Context context, @NonNull Source source) {
        if (context == null) {
            return new ArrayList<>();
        }
        if (source == Source.UPVOTED) {
            return Utils.loadUpvoted(context, true);
        }
        return Utils.loadFavorites(context, true);
    }

    static boolean idsMatchCache(@NonNull Context context,
                                 @NonNull Source source,
                                 @NonNull Snapshot snapshot) {
        ArrayList<Bookmark> cachedItems = loadCache(context, source);
        Set<Integer> cachedCommentIds = loadCommentIds(context, source);

        if (cachedItems.size() != snapshot.itemIds.size()
                || !cachedCommentIds.equals(snapshot.commentIds)) {
            return false;
        }

        for (int i = 0; i < cachedItems.size(); i++) {
            if (cachedItems.get(i).id != snapshot.itemIds.get(i)) {
                return false;
            }
        }

        return true;
    }

    static void saveIds(@NonNull Context context,
                        @NonNull Source source,
                        @NonNull Snapshot snapshot) {
        if (source == Source.UPVOTED) {
            Utils.saveUpvotedIds(context, snapshot.itemIds);
            Utils.saveUpvotedCommentIds(context, snapshot.commentIds);
        } else {
            Utils.saveFavoriteIds(context, snapshot.itemIds);
            Utils.saveFavoriteCommentIds(context, snapshot.commentIds);
        }
    }

    @NonNull
    private static Set<Integer> loadCommentIds(@NonNull Context context, @NonNull Source source) {
        if (source == Source.UPVOTED) {
            return Utils.loadUpvotedCommentIds(context);
        }
        return Utils.loadFavoriteCommentIds(context);
    }

    @NonNull
    private static ArrayList<Integer> normalizeItemIds(@NonNull List<Integer> itemIds) {
        ArrayList<Integer> normalizedItemIds = new ArrayList<>();
        for (int id : itemIds) {
            if (!normalizedItemIds.contains(id)) {
                normalizedItemIds.add(id);
            }
        }

        Collections.sort(normalizedItemIds, (id1, id2) -> Integer.compare(id2, id1));
        return normalizedItemIds;
    }

    @NonNull
    private static Set<Integer> normalizeCommentIds(@NonNull List<Integer> itemIds,
                                                    @NonNull List<Integer> commentIds) {
        Set<Integer> itemIdSet = new HashSet<>(itemIds);
        Set<Integer> normalizedCommentIds = new HashSet<>();
        for (int id : commentIds) {
            if (itemIdSet.contains(id)) {
                normalizedCommentIds.add(id);
            }
        }
        return normalizedCommentIds;
    }
}

File: app/src/main/java/com/simon/harmonichackernews/UserItemListRepository.kt
package com.simon.harmonichackernews

import android.content.Context
import com.simon.harmonichackernews.data.Bookmark
import com.simon.harmonichackernews.data.SavedItemCodec
import com.simon.harmonichackernews.data.SavedItemSnapshot
import com.simon.harmonichackernews.data.SavedItemSnapshots
import com.simon.harmonichackernews.utils.Utils

internal object UserItemListRepository {
    fun normalizeSnapshot(
        itemIds: List<Int>,
        commentIds: List<Int>
    ): Snapshot {
        return SavedItemSnapshots.normalize(itemIds, commentIds).toAndroidSnapshot()
    }

    fun loadCachedSnapshot(context: Context?, source: Source): Snapshot {
        if (context == null) return Snapshot(emptyList(), emptySet())

        val itemIds = loadCache(context, source)
            .map(Bookmark::id)
            .distinct()
            .sortedDescending()
        return Snapshot(itemIds, loadCommentIds(context, source))
    }

    fun loadCache(context: Context?, source: Source): ArrayList<Bookmark> {
        if (context == null) {
            return ArrayList()
        }
        return when (source) {
            Source.UPVOTED -> Utils.loadUpvoted(context, true)
            Source.FAVORITES -> Utils.loadFavorites(context, true)
        }
    }

    fun idsMatchCache(
        context: Context,
        source: Source,
        snapshot: Snapshot
    ): Boolean {
        val cachedItems = loadCache(context, source)
        val cachedCommentIds = loadCommentIds(context, source)

        return snapshot.toShared().matches(
            SavedItemCodec.fromBookmarks(cachedItems),
            cachedCommentIds,
        )
    }

    fun saveIds(
        context: Context,
        source: Source,
        snapshot: Snapshot
    ) {
        when (source) {
            Source.UPVOTED -> {
                Utils.saveUpvotedIds(context, snapshot.itemIds)
                Utils.saveUpvotedCommentIds(context, snapshot.commentIds)
            }
            Source.FAVORITES -> {
                Utils.saveFavoriteIds(context, snapshot.itemIds)
                Utils.saveFavoriteCommentIds(context, snapshot.commentIds)
            }
        }
    }

    private fun loadCommentIds(context: Context, source: Source): Set<Int> =
        when (source) {
            Source.UPVOTED -> Utils.loadUpvotedCommentIds(context)
            Source.FAVORITES -> Utils.loadFavoriteCommentIds(context)
        }

    internal enum class Source {
        FAVORITES,
        UPVOTED
    }

    internal data class Snapshot(val itemIds: List<Int>, val commentIds: Set<Int>) {
        fun toShared(): SavedItemSnapshot = SavedItemSnapshot(itemIds, commentIds)
    }

    private fun SavedItemSnapshot.toAndroidSnapshot(): Snapshot = Snapshot(itemIds, commentIds)
}
